src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -980,6 +972,17 @@ private void InheritNullableStateOfMember(int targetContainerSlot, int valueCont
             }
         }
 
+        private void SetStateAndTrackForFinally(ref LocalState state, int slot, NullableFlowState newState)
+        {
+            state[slot] = newState;
+            if (newState == NullableFlowState.MaybeNull && _tryState.HasValue)
```

Comment:
I'm not very familiar with the nullable walker yet, but here's my general question:

For non-monotonic transfer we need to really track state transitions in two places,

1. The general state
2. The state from ReachableBottom

so that those states can be merged at the appropriate point. What this code seems to be
doing is handling the setting of both states as long as we're in a situation where we may
have a Meet.

What I'm not sure about is why we need to manually set the try state. Won't it be updated
by Joining with the full state at the end of the try?

Maybe more importantly, I'm worried that this deals with the try-finally case, but there's no
generalization for non-monotonic transfer.

Reply:
The try statement cannot be handled just by a Meet or Join operation at the end of the try block; we also need to be sensitive to the order of operations

``` c#
string s;
try
{
    s = string.Empty;
}
finally
{
    s = null;
}
s.ToString(); // would be OK if we use a Meet of the state from the try block and the state from the finally block
```

Instead, we keep track of the mutations to null taken in the finally block using the field `_tryState`.  See the function `AbstractFlowPass.VisitFinallyBlockWithAnyTransferFunction`.  We Meet that with the state at the end of the finally block to get only those assignments that reached the end of the finally block.  Note that this is only a heuristic, and it can be fooled!  But nobody is likely to notice.  Here is one way to fool it:

``` c#
    public static void M(bool b)
    {
        string? s = null;
        try
        {
            s = string.Empty;
        }
        finally
        {
            if (b)
            {
                s = null;
                throw null!;
            }
        }

        s.ToString(); // warning?
    }
```

The generalization for non-monotonic transfer is already implemented in the base walker.  This is just responsible for wiring things together for the try statement.  Basically, any derived walker that performs mutations of the state must also Join that mutation into _tryState.  As an optimization, when the state (per variable) is so simple as it is here or in definite assignment, we can simplify that to only modify _tryState when we are making a change toward the top of the lattice - i.e. changing a variable to MaybeNull (or NotAssigned).  That is why we just assign to that one slot when the new state is MaybeNull. 

If that doesn't confuse you into submission I don't know what will ;)


---
In reply to: [263959663](https://github.com/dotnet/roslyn/pull/33962#discussion_r263959663) [](ancestors = 263959663)